pq-crypto/s2n_pq.c

```diff
@@ -82,19 +93,62 @@ bool s2n_cpu_supports_sikep434r2_asm() {
 #endif /* defined(S2N_SIKEP434R2_ASM) */
 }
 
+void s2n_bike_r3_x86_64_opt_init() {
+#if defined(S2N_BIKE_R3_X86_64)
+
+    uint32_t eax, ebx, ecx, edx;
+    if (!s2n_get_cpuid_count(EXTENDED_FEATURES_LEAF, EXTENDED_FEATURES_SUBLEAF_ZERO, &eax, &ebx, &ecx, &edx)) {
+        return;
+    }
+
+    bike_r3_avx2_enabled    = (ebx & EBX_BIT_AVX2) != 0;
+    bike_r3_avx512_enabled  = (ebx & EBX_BIT_AVX512) != 0;
+    bike_r3_pclmul_enabled  = bike_r3_avx2_enabled;
+    bike_r3_vpclmul_enabled = (ecx & ECX_BIT_VPCLMUL) != 0;
+#else
+    bike_r3_avx2_enabled    = false;
+    bike_r3_avx512_enabled  = false;
+    bike_r3_pclmul_enabled  = false;
+    bike_r3_vpclmul_enabled = false;
+#endif
+}
+
 #else /* defined(S2N_CPUID_AVAILABLE) */
 
 /* If CPUID is not available, we cannot perform necessary run-time checks. */
 bool s2n_cpu_supports_sikep434r2_asm() {
     return false;
 }
 
+void s2n_bike_r3_x86_64_opt_init() {
+    bike_r3_avx2_enabled    = false;
+    bike_r3_avx512_enabled  = false;
+    bike_r3_pclmul_enabled  = false;
+    bike_r3_vpclmul_enabled = false;
+}
+
 #endif /* defined(S2N_CPUID_AVAILABLE) */
 
 bool s2n_sikep434r2_asm_is_enabled() {
     return sikep434r2_asm_enabled;
 }
 
+bool s2n_bike_r3_is_avx2_enabled() {
+    return bike_r3_avx2_enabled;
+}
+
+bool s2n_bike_r3_is_avx512_enabled() {
+    return bike_r3_avx512_enabled;
+}
+
+bool s2n_bike_r3_is_pclmul_enabled() {
+    return bike_r3_pclmul_enabled;
+}
+
+bool s2n_bike_r3_is_vpclmul_enabled() {
+    return bike_r3_vpclmul_enabled;
+}
```

Comment:
In addition the these specific queries, it's probably useful to have a consolidated query like

```
bool s2n_bikel1r3_optimizations_are_enabled() {
    return bike_r3_avx2_enabled || bike_r3_avx512_enabled || bike_r3_pclmul_enabled || bike_r3_vpclmul_enabled;
}
```

In unit/KAT tests specifically, we want to know if we need to run the test twice - once with the portable implementation, and once if _any_ of the optimizations are enabled.

Reply:
I added new unit tests to cover all the optimizations, let me know if that's ok.